Add new index and cluster level settings to limit the total primary shards per node and per index

pandeydivyansh1803 opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
For remote store backed cluster, Segment Replication is used as the replication strategy. With segment replication, segments are created only on primary shard and these segments are copied to the replica shards. As segment creation is CPU intensive, we have observed CPU skew between nodes of the same cluster where primary shards are not balanced.

The earlier attempts to rebalance primary shards across nodes (https://github.com/opensearch-project/OpenSearch/pull/6422, https://github.com/opensearch-project/OpenSearch/issues/12250) are definitely helping to reduce the skew but they work on the best effort basis and don’t add any constraint.

Implement new setting in OpenSearch:
`index.routing.allocation.total_primary_shards_per_node`: An index-level setting to limit primary shards per node for a specific index. Store this limit (indexTotalPrimaryShardsPerNodeLimit) in index metadata, similar to indexTotalShardsPerNodeLimit.
`cluster.routing.allocation.total_primary_shards_per_node`: A cluster-level setting to limit the total primary shards on a node.

These settings will enhance control over primary shard distribution, improving cluster balance and performance management.
The existing ShardsLimitAllocationDecider class already contains the necessary infrastructure and logic to evaluate shard allocation constraints. It has access to the current cluster state, routing information, and methods to check shard counts per node. Given this existing functionality, we propose implementing the two new primary shard limit settings within this class. This approach leverages the current decision-making framework, ensuring consistency with existing allocation rules and minimizing code duplication. By extending the ShardsLimitAllocationDecider, we can efficiently integrate the new primary shard limit checks into the existing allocation decision process.

### Related Issues
Resolves #17293
<!-- List any other related issues here -->

### Check List
- [✔️] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [✔️] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/9301), [https://github.com/opensearch-project/documentation-website/issues/9301)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

cwperks commented on untitled/.gitignore:
```diff
@@ -0,0 +1,29 @@
+### IntelliJ IDEA ###
```
Looks like the files under the `untitled/` folder got added to this PR inadvertantly. Can you remove this file and the one below?

pandeydivyansh1803 replied:
Done. Removed the unintended files from the PR.

pandeydivyansh1803 commented on server/src/main/java/org/opensearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -115,7 +130,29 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
+    /**
+     * Checks whether the given shardRouting's index uses segment replication
+     */
+    private boolean isIndexSegmentReplicationUsed(RoutingAllocation allocation, ShardRouting shardRouting) {
```
acknowledged and resolved

pandeydivyansh1803 commented on .gitignore:
acknowledged and resolved

pandeydivyansh1803 commented on server/src/main/java/org/opensearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -115,7 +130,29 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
+    /**
+     * Counts the number of primary shards on a given node for a given index
+     */
+    private int countIndexPrimaryShards(RoutingNode node, ShardRouting shardRouting) {
```
acknowledged and resolved

astute-decipher reviewed (approved):
Changes LGTM

shourya035 commented on server/src/main/java/org/opensearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -115,7 +149,15 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
+    private boolean isSegRepEnabledIndex(RoutingAllocation allocation, ShardRouting shardRouting) {
+        IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
+        Settings indexSettings = indexMetadata.getSettings();
+        return IndexMetadata.INDEX_REPLICATION_TYPE_SETTING.get(indexSettings) == ReplicationType.SEGMENT;
```
We might need to handle dual replication mode also. During the remote store migration process, an index behaves like a pseudo segment replication backed index, but don't update the index setting unless it is completely moved over to the remote store enabled nodes. [Reference](https://github.com/opensearch-project/OpenSearch/issues/13252)

There should be a method named `indexSettings.assignedOnRemoteNode()` which can be used as an `&&` condition to this\

pandeydivyansh1803 replied:
We're removing this check altogether

shourya035 commented on server/src/main/java/org/opensearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -125,11 +167,14 @@ private Decision doDecide(
         BiPredicate<Integer, Integer> decider
     ) {
         final int indexShardLimit = allocation.metadata().getIndexSafe(shardRouting.index()).getIndexTotalShardsPerNodeLimit();
+        final int indexPrimaryShardLimit = allocation.metadata()
```
We can decouple fetching `allocation.metadata().getIndexSafe(shardRouting.index())` to a separate variable and re-use that. Reduces redundant code

shourya035 commented on server/src/internalClusterTest/java/org/opensearch/cluster/routing/allocation/decider/ShardsLimitAllocationDeciderIT.java:
```diff
@@ -0,0 +1,512 @@
+
+    }
+
+    public void testIndexSpecificShardLimit() {
```
We are enforcing the limit to segrep backed indices only. But I don't see the `INDEX_REPLICATION_TYPE_SETTING` setting being added during index creation here. Can we have that assertion in place? Fetch the indices created in this test, and assert that the `INDEX_REPLICATION_TYPE_SETTING` for all of them is set to `SEGMENT`

shourya035 commented:
The cluster level setting `cluster.routing.allocation.total_primary_shards_per_node` is applied regardless of whether the cluster is hosting `DOCUMENT` or `SEGMENT` replication type indices. Today, we allow the end-user to have a cluster with indices having mix of both replication types. 

Did we do some benchmarks on this? Maybe by creating 20% SEGMENT replication type indices and the rest 80% on DOCUMENT replication type indices. I am worried that not having the index level setting work on DOCUMENT replication based indices would result in triggering a lot of rebalances. Too many rebalances would also have a performance overhead.

pandeydivyansh1803 commented:
We’ve acknowledged the fact that in open-source code it is possible that some index/indices in a cluster are using document replication whereas other are using segment replication. In few cases this config combined with using `cluster.routing.allocation.total_primary_shards_per_node` may result in an imbalance even after balancing primaries on account of the fact that the replicas remaining are not all identical in terms of compute. Replicas of DOCrep enabled indices require more compute when compared to replicas of SEGrep enabled indices. At cluster level we’ve moved ahead and implemented this setting assuming that such scenarios are very seldom. It makes more sense to keep the index level setting consistent with the cluster level setting moreover, this doesn’t have any explicit disadvantage to let end user balance the total primaries on a node at the index level for both SEGrep and DOCrep enabled indices. Moving ahead with removing the check added in previous commits to limit index level settings only for SEGrep enabled indices, allowing the `index.routing.allocation.total_primary_shards_per_node` to take effect for both DOCrep and SEGrep enabled indices.

shourya035 commented on server/src/main/java/org/opensearch/action/admin/cluster/settings/TransportClusterUpdateSettingsAction.java:
```diff
@@ -324,4 +333,24 @@ private void validateAllNodesOfSameType(DiscoveryNodes discoveryNodes) {
+
+        if (!allNodesRemoteStoreEnabled) {
+            throw new IllegalArgumentException(
+                "Setting [" + CLUSTER_TOTAL_PRIMARY_SHARDS_PER_NODE_SETTING.getKey() + "] can only be used with remote store domains"
```
> can only be used with remote store domains

Change this to 

```
can only be used with remote store enabled clusters
```

pandeydivyansh1803 replied:
noted and changed.

shourya035 commented on server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:
```diff
@@ -1844,6 +1846,27 @@ public static void validateRefreshIntervalSettings(Settings requestSettings, Clu
+        boolean isRemoteStoreEnabled = IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.get(indexSettings);
+        if (!isRemoteStoreEnabled) {
+            throw new IllegalArgumentException(
+                "Setting [" + INDEX_TOTAL_PRIMARY_SHARDS_PER_NODE_SETTING.getKey() + "] can only be used with remote store domains"
```
Same here. Change this to 

```
can only be used with remote store enabled clusters
```

pandeydivyansh1803 replied:
noted and changed.

shourya035 commented on server/src/main/java/org/opensearch/action/admin/cluster/settings/TransportClusterUpdateSettingsAction.java:
```diff
@@ -257,6 +258,14 @@ public void onFailure(String source, Exception e) {
+                        Settings settings = Settings.builder().put(request.transientSettings()).put(request.persistentSettings()).build();
+
+                        int newValue = CLUSTER_TOTAL_PRIMARY_SHARDS_PER_NODE_SETTING.get(settings);
+                        validateClusterTotalPrimaryShardsPerNodeSetting(currentState, newValue);
```
Let's move this entire logic within the `validateClusterTotalPrimaryShardsPerNodeSetting` method. Bail out, if the pre-requisites are not met.

pandeydivyansh1803 replied:
resolved.

sachinpkale commented:
PR title should follow imperative style. When we merge the PR, the PR title becomes commit message heading.
Instead of `Added`, we should use `Add`

sachinpkale commented on server/src/main/java/org/opensearch/cluster/routing/RoutingNode.java:
```diff
@@ -384,6 +384,16 @@ public int numberOfOwningShards() {
         return shards.size() - relocatingShards.size();
     }
 
+    public int numberOfOwningPrimaryShards() {
```
Is there any direct way to get this number instead of iterating over all the shards? If the node has 1000s of shards, can this create an issue?

pandeydivyansh1803 replied:
Considering the scenarios where a given node has large number of shards, iterating through all the shards will contribute to latency during decision making. `RoutingNode` class already maintains a `LinkedHashMap` to keep track of primary shards. We can marginally optimise on the time complexity by iterating through this and eliminating primary shards which are in `RELOCATING` state. Alternate approach could be to dynamically keep track of relocating primary shards on a node. The difference in the size of total primary shards and relocating primary shards on a node would give the required value in constant time but this has an added space complexity of maintaining all the relocating primary shards.

sachinpkale commented on server/src/main/java/org/opensearch/cluster/routing/RoutingNode.java:
```diff
@@ -144,15 +200,17 @@ public RoutingNode(String nodeId, DiscoveryNode node, ShardRouting... shardRouti
-        this.relocatingShards = new LinkedHashSet<>();
+        final LinkedHashSet<ShardRouting> relocatingShards = new LinkedHashSet<>();
+        final LinkedHashSet<ShardRouting> relocatingPrimaryShards = new LinkedHashSet<>();
+        this.relocatingShardsBucket = new RelocatingShardsBucket(relocatingShards, relocatingPrimaryShards);
```
If we are always passing empty hashset in the constructor, do we need to pass anything at all?

pandeydivyansh1803 replied:
I can use an empty constructor here, will update it.

linuxpi commented:
Please resolve conflicts in `CHANGELOG-3.0.md` and ensure gradle check is passing

linuxpi commented on server/src/main/java/org/opensearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -84,7 +87,19 @@ public class ShardsLimitAllocationDecider extends AllocationDecider {
+     * node for segment replication enabled indices. Negative values are interpreted as unlimited.
+     */
+    public static final Setting<Integer> INDEX_TOTAL_PRIMARY_SHARDS_PER_NODE_SETTING = Setting.intSetting(
+        "index.routing.allocation.total_primary_shards_per_node",
```
we would need a corresponding documentation update PR as well.

pandeydivyansh1803 replied:
Create documentation issue and added link in the description

linuxpi commented on server/src/main/java/org/opensearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -70,6 +72,7 @@ public class ShardsLimitAllocationDecider extends AllocationDecider {
     public static final String NAME = "shards_limit";
 
     private volatile int clusterShardLimit;
+    private volatile int clusterPrimaryShardLimit;
```
nit: `clusterPrimaryShardPerNodeLimit`

pandeydivyansh1803 replied:
`clusterPrimaryShardLimit` is used to keep it consistent with already used `clusterShardLimit`. If still required the name can be updated.

linuxpi commented on server/src/main/java/org/opensearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java:
```diff
@@ -165,6 +212,20 @@ private Decision doDecide(
+        }
         return allocation.decision(
             Decision.YES,
             NAME,
```
lets update the text here to include the new settings

pandeydivyansh1803 replied:
Done

pandeydivyansh1803 commented on server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:
```diff
@@ -1844,6 +1846,27 @@ public static void validateRefreshIntervalSettings(Settings requestSettings, Clu
+
+        // Check if remote store is enabled
+        boolean isRemoteStoreEnabled = IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.get(indexSettings);
+        if (!isRemoteStoreEnabled) {
```
Initially this setting was being implemented for mixed mode clusters but further discussions aligned implementing it only for remote store enabled clusters as of now and later extend it to mixed mode cluster on need basis. The reason was that when cluster is remote store enabled segment replication is used and replicas don't perform CPU intensive segment and hence need to balance primaries in this case.

pandeydivyansh1803 replied:
Will a todo for SegRep to DocRep migration case handling.

pandeydivyansh1803 commented on CHANGELOG-3.0.md:
```diff
@@ -16,6 +16,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Views, simplify data access and manipulation by providing a virtual layer over one or more indices ([#11957](https://github.com/opensearch-project/OpenSearch/pull/11957))
 - Added pull-based Ingestion (APIs, for ingestion source, a Kafka plugin, and IngestionEngine that pulls data from the ingestion source) ([#16958](https://github.com/opensearch-project/OpenSearch/pull/16958))
 - Added ConfigurationUtils to core for the ease of configuration parsing [#17223](https://github.com/opensearch-project/OpenSearch/pull/17223)
+- Add cluster and index level settings to limit the total primary shards per node and per index [#17295](https://github.com/opensearch-project/OpenSearch/pull/17295)
```
Yes this scenario will create a red cluster as 10 primaries will remain unassigned. User needs to carefully tune the values of setting.

linuxpi commented:
@pandeydivyansh1803 Changes LGTM. Please add some tests to cover the cases where user is not able to set the new index/cluster settings for non remote store cluster step.

pandeydivyansh1803 commented on server/src/internalClusterTest/java/org/opensearch/cluster/routing/allocation/decider/ShardsLimitAllocationDeciderRemoteStoreEnabledIT.java:
```diff
@@ -0,0 +1,248 @@
+        // Create first index with primary shard limit
+        Settings firstIndexSettings = Settings.builder()
+            .put(remoteStoreIndexSettings(0, 4))  // 4 shards, 0 replicas
+            .put(INDEX_TOTAL_PRIMARY_SHARDS_PER_NODE_SETTING.getKey(), 1)
```
Added relevant integration tests for the same.

linuxpi commented:
@pandeydivyansh1803 can you manually raise the backport PR?
